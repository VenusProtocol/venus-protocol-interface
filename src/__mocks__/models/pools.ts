import { Pool } from 'types';

import { assetData } from '__mocks__/models/asset';

export const poolData: Pool[] = [
  {
    id: '1',
    assets: assetData,
    name: 'Venus',
    riskLevel: 'MINIMAL',
    description: 'Fake description 1',
    isIsolated: false,
  },
  {
    id: '2',
    assets: assetData,
    name: 'Metaverse',
    riskLevel: 'VERY_HIGH',
    description: 'Fake description 2',
    isIsolated: true,
  },
  {
    id: '3',
    assets: assetData,
    name: 'Gaming',
    riskLevel: 'MEDIUM',
    description: 'Fake description 3',
    isIsolated: true,
  },
];
