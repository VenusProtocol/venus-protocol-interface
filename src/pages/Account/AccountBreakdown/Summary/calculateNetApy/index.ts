import BigNumber from 'bignumber.js';

const calculateNetApy = ({
  supplyBalanceCents,
  yearlyEarningsCents,
}: {
  supplyBalanceCents: BigNumber;
  yearlyEarningsCents: BigNumber;
}) => {
  // Calculate net APY as a percentage of supply balance, based on yearly interests
  let netApyPercentage: number | undefined;
  if (supplyBalanceCents?.isEqualTo(0)) {
    netApyPercentage = 0;
  } else if (supplyBalanceCents && yearlyEarningsCents) {
    netApyPercentage = +yearlyEarningsCents
      .multipliedBy(100)
      .dividedBy(supplyBalanceCents)
      .toFixed(2);
  }
  return netApyPercentage;
};

export default calculateNetApy;
