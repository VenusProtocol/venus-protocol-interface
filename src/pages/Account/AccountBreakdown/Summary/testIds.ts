const TEST_IDS = {
  container: 'account-summary-container',
  stats: 'account-summary-stats',
  accountHealth: 'account-summary-account-health',
};

export default TEST_IDS;
