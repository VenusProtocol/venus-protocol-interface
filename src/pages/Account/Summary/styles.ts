import { css } from '@emotion/react';
import { useTheme } from '@mui/material';

export const useStyles = () => {
  const theme = useTheme();

  return {
    container: css`
      display: flex;
      justify-content: space-between;

      ${theme.breakpoints.down('xl')} {
        display: block;
        padding: 0;
        background-color: transparent;
      }
    `,
    getNetApyColor: ({ netApyPercentage }: { netApyPercentage: number }) =>
      netApyPercentage > 0 ? theme.palette.interactive.success : theme.palette.interactive.error,
    cellGroup: css`
      padding: 0;

      ${theme.breakpoints.down('xl')} {
        margin-bottom: ${theme.spacing(2)};
      }
    `,
    accountHealth: css`
      ${theme.breakpoints.down('xl')} {
        padding: ${theme.spacing(4)};
        background-color: ${theme.palette.background.paper};
        border-radius: ${theme.shape.borderRadius.large}px;
      }
    `,
    accountHealthProgressBar: css`
      margin-bottom: ${theme.spacing(3)};
    `,
    accountHealthFooter: css`
      display: flex;
      justify-content: flex-end;
      align-items: center;
    `,
    tooltip: css`
      display: flex;
    `,
    infoIcon: css`
      cursor: help;
    `,
    inlineLabel: css`
      margin-right: ${theme.spacing(1)};
    `,
    progressBar: css`
      margin-bottom: ${theme.spacing(3)};
    `,
    shieldIcon: css`
      margin-right: ${theme.spacing(2)};
    `,
    safeLimit: css`
      margin-right: ${theme.spacing(2)};
    `,
  };
};
