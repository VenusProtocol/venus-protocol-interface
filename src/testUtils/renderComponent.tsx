import React from 'react';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import { Web3Wrapper } from 'clients/web3';
import { AuthContext, IAuthContextValue } from 'context/AuthContext';
import { SuccessfulTransactionModalProvider } from 'context/SuccessfulTransactionModalContext';
import Theme from 'theme';
import { RefreshContextProvider } from 'context/RefreshContext';
import { VaiContext, IVaiContextValue } from 'context/VaiContext';
import { MuiThemeProvider } from 'theme/MuiThemeProvider/MuiThemeProvider';
import BigNumber from 'bignumber.js';

const renderComponent = (
  children: React.ReactElement | (() => React.ReactElement),
  {
    authContextValue = {},
    vaiContextValue = {},
  }: {
    authContextValue?: Partial<IAuthContextValue>;
    vaiContextValue?: Partial<IVaiContextValue>;
  } = {},
) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
      },
    },
  });
  const defaultAuthContextValues = {
    login: jest.fn(),
    logOut: jest.fn(),
    openAuthModal: jest.fn(),
    closeAuthModal: jest.fn(),
    account: undefined,
    ...authContextValue,
  };
  const defaultVaiContextValues = {
    userVaiMinted: new BigNumber(0),
    userVaiBalance: new BigNumber(0),
    userVaiEnabled: false,
    mintableVai: new BigNumber(0),
    ...vaiContextValue,
  };

  const renderRes = render(
    <Theme>
      <Web3Wrapper>
        <QueryClientProvider client={queryClient}>
          <MuiThemeProvider>
            <AuthContext.Provider value={defaultAuthContextValues}>
              <RefreshContextProvider>
                <VaiContext.Provider value={defaultVaiContextValues}>
                  <SuccessfulTransactionModalProvider>
                    <BrowserRouter>
                      <ToastContainer />

                      <Switch>
                        <Route
                          path="/"
                          component={typeof children === 'function' ? children : () => children}
                        />
                      </Switch>
                    </BrowserRouter>
                  </SuccessfulTransactionModalProvider>
                </VaiContext.Provider>
              </RefreshContextProvider>
            </AuthContext.Provider>
          </MuiThemeProvider>
        </QueryClientProvider>
      </Web3Wrapper>
    </Theme>,
  );

  return {
    ...renderRes,
    queryClient,
  };
};

export default renderComponent;
