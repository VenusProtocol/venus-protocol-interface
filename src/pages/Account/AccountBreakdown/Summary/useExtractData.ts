import BigNumber from 'bignumber.js';
import { useMemo } from 'react';

import { SAFE_BORROW_LIMIT_PERCENTAGE } from 'constants/safeBorrowLimitPercentage';
import { Pool, Vault } from 'types';
import {
  calculateDailyEarningsCents,
  calculateYearlyEarningsForAssets,
  calculateYearlyInterests,
  convertMantissaToTokens,
  formatCentsToReadableValue,
  formatPercentageToReadableValue,
} from 'utilities';

import calculateNetApy from './calculateNetApy';

interface UseExtractDataInput {
  pools: Pool[];
  vaults: Vault[];
  xvsPriceCents: BigNumber;
  vaiPriceCents: BigNumber;
}

const useExtractData = ({ pools, vaults, xvsPriceCents, vaiPriceCents }: UseExtractDataInput) =>
  useMemo(() => {
    const { totalBorrowCents, totalSupplyCents, borrowLimitCents } = pools.reduce(
      (acc, pool) => ({
        totalBorrowCents: acc.totalBorrowCents.plus(pool.userBorrowBalanceCents || 0),
        totalSupplyCents: acc.totalSupplyCents.plus(pool.userSupplyBalanceCents || 0),
        borrowLimitCents: acc.borrowLimitCents.plus(pool.userBorrowLimitCents || 0),
      }),
      {
        totalSupplyCents: new BigNumber(0),
        totalBorrowCents: new BigNumber(0),
        borrowLimitCents: new BigNumber(0),
      },
    );

    const { totalVaultStakeCents, yearlyVaultEarningsCents } = vaults.reduce(
      (accTotalVaultStakeCents, vault) => {
        const vaultStakeCents = convertMantissaToTokens({
          value: new BigNumber(vault.userStakedMantissa || 0),
          token: vault.stakedToken,
        }).multipliedBy(vault.stakedToken.symbol === 'XVS' ? xvsPriceCents : vaiPriceCents);

        return {
          totalVaultStakeCents: accTotalVaultStakeCents.totalVaultStakeCents.plus(vaultStakeCents),
          yearlyVaultEarningsCents: accTotalVaultStakeCents.yearlyVaultEarningsCents.plus(
            calculateYearlyInterests({
              balance: vaultStakeCents,
              interestPercentage: new BigNumber(vault.stakingAprPercentage),
            }),
          ),
        };
      },
      { totalVaultStakeCents: new BigNumber(0), yearlyVaultEarningsCents: new BigNumber(0) },
    );

    const yearlyAssetEarningsCents = pools.reduce((acc, pool) => {
      const yearlyPoolAssetsEarningsCents = calculateYearlyEarningsForAssets({
        assets: pool.assets,
      });

      return acc.plus(yearlyPoolAssetsEarningsCents || 0);
    }, new BigNumber(0));

    const yearlyEarningsCents = yearlyAssetEarningsCents.plus(yearlyVaultEarningsCents);
    const dailyEarningsCentsTmp = calculateDailyEarningsCents(yearlyEarningsCents).toNumber();

    const netApyPercentageTmp =
      yearlyAssetEarningsCents &&
      calculateNetApy({
        supplyBalanceCents: totalSupplyCents.plus(totalVaultStakeCents),
        yearlyEarningsCents,
      });

    const safeBorrowLimitCentsTmp = borrowLimitCents.multipliedBy(
      SAFE_BORROW_LIMIT_PERCENTAGE / 100,
    );

    const readableSafeBorrowLimitTmp = formatCentsToReadableValue({
      value: safeBorrowLimitCentsTmp,
    });

    const safeBorrowLimitPercentageTmp = formatPercentageToReadableValue(
      safeBorrowLimitCentsTmp.multipliedBy(100).dividedBy(borrowLimitCents),
    );

    return {
      dailyEarningsCents: dailyEarningsCentsTmp,
      netApyPercentage: netApyPercentageTmp,
      readableSafeBorrowLimit: readableSafeBorrowLimitTmp,
      safeBorrowLimitPercentage: safeBorrowLimitPercentageTmp,
      totalVaultStakeCents,
      totalBorrowCents,
      totalSupplyCents,
      borrowLimitCents,
    };
  }, [pools, vaults, xvsPriceCents, vaiPriceCents]);

export default useExtractData;
