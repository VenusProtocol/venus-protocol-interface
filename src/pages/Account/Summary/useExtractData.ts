import BigNumber from 'bignumber.js';
import { useMemo } from 'react';
import { Asset, Pool, Vault } from 'types';
import {
  areTokensEqual,
  calculateDailyEarningsCents,
  calculateYearlyEarningsForAssets,
  calculateYearlyInterests,
  convertWeiToTokens,
  formatCentsToReadableValue,
  formatToReadablePercentage,
} from 'utilities';

import { SAFE_BORROW_LIMIT_PERCENTAGE } from 'constants/safeBorrowLimitPercentage';
import { TOKENS } from 'constants/tokens';

import calculateNetApy from './calculateNetApy';

interface UseExtractDataInput {
  pools: Pool[];
  vaults: Vault[];
  xvsPriceCents: BigNumber;
  vaiPriceCents: BigNumber;
}

const useExtractData = ({ pools, vaults, xvsPriceCents, vaiPriceCents }: UseExtractDataInput) =>
  useMemo(() => {
    const { totalBorrowCents, totalSupplyCents, borrowLimitCents } = pools.reduce(
      (acc, pool) => ({
        totalBorrowCents: acc.totalBorrowCents.plus(pool.userBorrowBalanceCents || 0),
        totalSupplyCents: acc.totalSupplyCents.plus(pool.userSupplyBalanceCents || 0),
        borrowLimitCents: acc.borrowLimitCents.plus(pool.userBorrowLimitCents || 0),
      }),
      {
        totalSupplyCents: new BigNumber(0),
        totalBorrowCents: new BigNumber(0),
        borrowLimitCents: new BigNumber(0),
      },
    );

    const { totalVaultStakeCents, yearlyVaultEarningsCents } = vaults.reduce(
      (accTotalVaultStakeCents, vault) => {
        const vaultStakeCents = convertWeiToTokens({
          valueWei: new BigNumber(vault.userStakedWei || 0),
          token: vault.stakedToken,
        }).multipliedBy(
          areTokensEqual(vault.stakedToken, TOKENS.xvs) ? xvsPriceCents : vaiPriceCents,
        );

        return {
          totalVaultStakeCents: accTotalVaultStakeCents.totalVaultStakeCents.plus(vaultStakeCents),
          yearlyVaultEarningsCents: accTotalVaultStakeCents.yearlyVaultEarningsCents.plus(
            calculateYearlyInterests({
              balance: vaultStakeCents,
              interestPercentage: new BigNumber(vault.stakingAprPercentage),
            }),
          ),
        };
      },
      { totalVaultStakeCents: new BigNumber(0), yearlyVaultEarningsCents: new BigNumber(0) },
    );

    const assets = pools.reduce((acc, pool) => [...acc, ...pool.assets], [] as Asset[]);

    const yearlyAssetEarningsCents = new BigNumber(
      calculateYearlyEarningsForAssets({
        assets,
      }) || 0,
    );

    const yearlyEarningsCents = yearlyAssetEarningsCents.plus(yearlyVaultEarningsCents);
    const dailyEarningsCentsTmp = calculateDailyEarningsCents(yearlyEarningsCents).toNumber();

    const netApyPercentageTmp =
      yearlyAssetEarningsCents &&
      calculateNetApy({
        supplyBalanceCents: totalSupplyCents.plus(totalVaultStakeCents),
        yearlyEarningsCents,
      });

    const safeBorrowLimitCentsTmp = borrowLimitCents.multipliedBy(
      SAFE_BORROW_LIMIT_PERCENTAGE / 100,
    );

    const readableSafeBorrowLimitTmp = formatCentsToReadableValue({
      value: safeBorrowLimitCentsTmp,
    });

    const safeBorrowLimitPercentageTmp = formatToReadablePercentage(
      safeBorrowLimitCentsTmp.multipliedBy(100).dividedBy(borrowLimitCents),
    );

    return {
      dailyEarningsCents: dailyEarningsCentsTmp,
      netApyPercentage: netApyPercentageTmp,
      readableSafeBorrowLimit: readableSafeBorrowLimitTmp,
      safeBorrowLimitPercentage: safeBorrowLimitPercentageTmp,
      totalVaultStakeCents,
      totalBorrowCents,
      totalSupplyCents,
      borrowLimitCents,
    };
  }, [pools, vaults, xvsPriceCents, vaiPriceCents]);

export default useExtractData;
