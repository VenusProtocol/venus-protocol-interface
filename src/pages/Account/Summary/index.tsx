/** @jsxImportSource @emotion/react */
import { Paper, Typography } from '@mui/material';
import { BorrowLimitUsedAccountHealth, Cell, CellGroup, Icon, Tooltip } from 'components';
import React from 'react';
import { useTranslation } from 'translation';
import { Asset } from 'types';
import { formatCentsToReadableValue, formatToReadablePercentage } from 'utilities';

import { useStyles } from './styles';
import TEST_IDS from './testIds';
import useExtractData from './useExtractData';

export interface SummaryProps {
  assets: Asset[];
  safeBorrowLimitPercentage?: number;
  displayAccountHealth?: boolean;
  className?: string;
}

export const Summary: React.FC<SummaryProps> = ({
  assets,
  displayAccountHealth,
  safeBorrowLimitPercentage,
  className,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const {
    totalSupplyCents,
    totalBorrowCents,
    borrowLimitCents,
    readableSafeBorrowLimit,
    dailyEarningsCents,
    netApyPercentage,
  } = useExtractData({
    assets,
    safeBorrowLimitPercentage,
  });

  const cells: Cell[] = [
    {
      label: t('account.marketBreakdown.cellGroup.netApy'),
      value: formatToReadablePercentage(netApyPercentage),
      tooltip: t('account.marketBreakdown.cellGroup.netApyTooltip'),
      color: styles.getNetApyColor({ netApyPercentage: netApyPercentage || 0 }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.dailyEarnings'),
      value: formatCentsToReadableValue({ value: dailyEarningsCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalSupply'),
      value: formatCentsToReadableValue({ value: totalSupplyCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalBorrow'),
      value: formatCentsToReadableValue({ value: totalBorrowCents }),
    },
  ];

  return (
    <Paper css={styles.container} className={className} data-testid={TEST_IDS.container}>
      <CellGroup
        smallValues={displayAccountHealth}
        cells={cells}
        css={styles.cellGroup}
        data-testid={TEST_IDS.stats}
      />

      {displayAccountHealth && safeBorrowLimitPercentage && (
        <div css={styles.accountHealth} data-testid={TEST_IDS.accountHealth}>
          <BorrowLimitUsedAccountHealth
            variant="borrowLimitUsed"
            borrowBalanceCents={totalBorrowCents.toNumber()}
            borrowLimitCents={borrowLimitCents.toNumber()}
            safeBorrowLimitPercentage={safeBorrowLimitPercentage}
            css={styles.accountHealthProgressBar}
          />

          <div css={styles.accountHealthFooter}>
            <Icon name="shield" css={styles.shieldIcon} />

            <Typography component="span" variant="small2" css={styles.inlineLabel}>
              {t('myAccount.safeLimit')}
            </Typography>

            <Typography
              component="span"
              variant="small1"
              color="text.primary"
              css={styles.safeLimit}
            >
              {readableSafeBorrowLimit}
            </Typography>

            <Tooltip
              css={styles.tooltip}
              title={t('myAccount.safeLimitTooltip', { safeBorrowLimitPercentage })}
            >
              <Icon css={styles.infoIcon} name="info" />
            </Tooltip>
          </div>
        </div>
      )}
    </Paper>
  );
};

export default Summary;
