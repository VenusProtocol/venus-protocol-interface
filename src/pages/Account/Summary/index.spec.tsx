import React from 'react';

import { poolData } from '__mocks__/models/pools';
import renderComponent from 'testUtils/renderComponent';

import Summary from '.';
import TEST_IDS from './testIds';

jest.mock('clients/api');

describe('pages/Account/Summary', () => {
  it('renders without crashing', () => {
    renderComponent(<Summary pools={poolData} />);
  });

  it('displays stats correctly', () => {
    const { getByTestId } = renderComponent(<Summary pools={poolData} />);

    expect(getByTestId(TEST_IDS.stats).textContent).toMatchSnapshot();
  });

  it('displays account health when passing displayAccountHealth prop as true', () => {
    const { getByTestId } = renderComponent(<Summary pools={poolData} displayAccountHealth />);

    expect(getByTestId(TEST_IDS.accountHealth).textContent).toMatchSnapshot();
  });
});
