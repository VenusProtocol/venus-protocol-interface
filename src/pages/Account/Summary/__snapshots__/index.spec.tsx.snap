// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pages/Account/Summary > displays account health when passing displayAccountHealth prop as true 1`] = `"Borrow limit used:6.4%Limit:$1.92K.Your safe limit:$1.54K"`;

exports[`pages/Account/Summary > displays stats correctly 1`] = `"Net APY0%Daily Earnings$0.05Total Supply$1.23MTotal Borrow$123.33"`;

exports[`pages/Account/Summary > displays total vault stake when passing vaults prop and displayTotalVaultStake prop as true 1`] = `"Net APY0%Daily Earnings$0.13Total Supply$1.23MTotal Borrow$123.33Total Vault Stake$233.00"`;
